api/src/database/migrations/20240321A-permissions-policies.ts

```diff
@@ -0,0 +1,173 @@
+import type { Knex } from 'knex';
+import { isEqual } from 'lodash-es';
+import { ItemsService } from '../../services/index.js';
+import { getSchema } from '../../utils/get-schema.js';
+
+interface OldPermission {
+	id: number;
+	role: string;
+	collection: string;
+	action: string;
+	permissions: null | Record<string, unknown>;
+	validation: null | Record<string, unknown>;
+	presets: null | Record<string, unknown>;
+	fields: null | string[];
+}
+
+interface NewPermission {
+	id: number;
+	policy: string;
+	collections: string[];
+	actions: string[];
+	type: string;
+	rule: null | Record<string, unknown> | string[];
+}
+
+const types = ['access', 'fields', 'validation', 'presets'] as const;
+
+function jsonValueExists(val: null | Record<string, unknown> | string[]): boolean {
+	try {
+		if (val === null) return false;
+
+		if (Array.isArray(val)) return val.length > 0;
+
+		return Object.keys(val).length !== 0;
+	} catch {
+		return false;
+	}
+}
+
+// NOTE: knex.schema.renameTable isn't reliable cross vendor, so can't be used here
+export async function up(knex: Knex): Promise<void> {
+	// Create policies table that's identical to roles
+	await knex.schema.createTable('directus_policies', (table) => {
+		table.uuid('id').primary();
+		table.string('name', 100).notNullable();
+		table.string('icon', 30).notNullable();
+		table.text('description');
+		table.text('ip_access');
+		table.boolean('enforce_tfa');
+		table.boolean('admin_access');
+		table.boolean('app_access');
+	});
+
+	// Copy existing roles 1-1 to policies
+	const roles = await knex.select('*').from('directus_roles');
+
+	if (roles.length > 0) {
+		await knex('directus_policies').insert(roles);
+	}
+
+	// Drop access control fields from roles
+	await knex.schema.alterTable('directus_roles', (table) => {
+		table.dropColumn('ip_access');
+		table.dropColumn('enforce_tfa');
+		table.dropColumn('admin_access');
+		table.dropColumn('app_access');
+	});
+
+	// Migrate existing permissions to new structure
+	// Gotta use the ItemsService to ensure JSON is cast back and forth to the DB format correctly
+	const permissions = (await new ItemsService('directus_permissions', {
+		knex: knex,
+		schema: await getSchema({ database: knex }),
+	}).readByQuery({
+		fields: ['id', 'role', 'collection', 'action', 'permissions', 'fields', 'validation', 'presets'],
+	})) as OldPermission[];
```

Comment:
We would need to add the `limit`.

```suggestion
	const permissions = (await new ItemsService('directus_permissions', {
		knex: knex,
		schema: await getSchema({ database: knex }),
	}).readByQuery({
		fields: ['id', 'role', 'collection', 'action', 'permissions', 'fields', 'validation', 'presets'],
		limit: -1,
	})) as OldPermission[];
```

On the contrary, as this is a migration, should we utilise the transformers from PayloadService instead of ItemsService as there _might_ be complications such as revisions and limits etc? 🤔 

https://github.com/directus/directus/blob/69479f4b629401a4985785c12103d6ad02de5f69/api/src/services/payload.ts#L59-L147

https://github.com/directus/directus/blob/69479f4b629401a4985785c12103d6ad02de5f69/api/src/services/payload.ts#L219-L244

Reply:
Good catch with the `limit`!

> On the contrary, as this is a migration, should we utilise the transformers from PayloadService instead of ItemsService as there might be complications such as revisions and limits etc? 🤔

What complication are you thinking of, besides the `limit`?

Reply:
The saving of revisions and activity might be unnecessary as it's a migration. 
Flows & extension hooks might get in the way, causing uncertainties which could be hard to triage or revert. 👀

Reply:
Ah gotcha, you mean specifically for insertion 👍 Extensions do not run, but I guess Flows could indeed conflict. Seems indeed better to avoid `ItemsService`.

Reply:
Reading too actually, as we have that items.read hook